### PR TITLE
feat: Define procedure related traits

### PR DESCRIPTION
**Pull request by evenyag**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

This PR ports the traits needed by the procedure framework from the [PoC branch](https://github.com/evenyag/greptimedb/tree/poc/procedure-api). The [RFC](https://github.com/GreptimeTeam/greptimedb/pull/836) explains the concept of `Procedure`, `ProcedureId`, and `ProcedureManager`.

Additionally, the `Procedure` trait has a `lock_key()` method that returns a `LockKey` struct to tell the `ProcedureManager` to acquire lock.

```rust
#[async_trait]
pub trait Procedure: Send + Sync {
    /// Type name of the procedure.
    fn type_name(&self) -> &str;

    // ......

    /// Returns the [LockKey] if this procedure needs to acquire lock.
    fn lock_key(&self) -> Option<LockKey>;
}
```

Note that `rollback` is not supported so the `Procedure` trait doesn't have the `rollback()` method.

The ProcedureManager also has a `recover()` to recover procedures from the storage layer and a `procedure_state()` method to query the procedure's status.

```rust
#[async_trait]
pub trait ProcedureManager: Send + Sync + 'static {
    // ......

    /// Recovers unfinished procedures and reruns them.
    ///
    /// Callers should ensure all loaders are registered.
    async fn recover(&self) -> Result<()>;

    /// Query the procedure state.
    ///
    /// Returns `Ok(None)` if the procedure doesn't exist.
    async fn procedure_state(&self, procedure_id: ProcedureId) -> Result<Option<ProcedureState>>;
}
```

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
- #286 
- #836

### CODE REVIEW

File: src/common/procedure/src/procedure.rs [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    /// Type name of the procedure.
+    fn type_name(&self) -> &str;
+
+    /// Execute the procedure.
+    ///
+    /// The implementation must be idempotent.
+    async fn execute(&mut self, ctx: &Context) -> Result<Status>;
+
+    /// Dump the state of the procedure to a string.
+    fn dump(&self) -> Result<String>;
```
Comment (MichaelScofield) [inline src/common/procedure/src/procedure.rs:77]: Maybe we can just make `Procedure` trait extend `Display` trait?

File: src/common/procedure/src/procedure.rs [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+pub struct ProcedureId(Uuid);
+
+impl ProcedureId {
+    /// Returns a new unique [ProcedureId] randomly.
+    pub fn random() -> ProcedureId {
+        ProcedureId(Uuid::new_v4())
+    }
+
+    /// Parses id from string.
+    pub fn parse_str(input: &str) -> Option<ProcedureId> {
```
Comment (MichaelScofield) [inline src/common/procedure/src/procedure.rs:133]: Why not return `Result<ProcedureId>`, to provide a more meaningful context?

**Review by MichaelScofield — APPROVED** [review]

File: src/common/procedure/src/procedure.rs [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+    /// Returns the [LockKey] if this procedure needs to acquire lock.
+    fn lock_key(&self) -> Option<LockKey>;
+}
+
+/// A key to identify the lock.
+// We might hold multiple keys in this struct. When there are multiple keys, we need to sort the
+// keys lock all the keys in order to avoid dead lock.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct LockKey(String);
```
[the remark this answers is not in the document]
Comment (evenyag) [reply src/common/procedure/src/procedure.rs:87]: Currently, they are just the full table name and don't have a uniform prefix.

File: src/common/procedure/src/procedure.rs (hunk 1 above)
Comment (evenyag) [reply src/common/procedure/src/procedure.rs:77]: This method behaves like `serialize()` and is different from `display`

**Review by fengjiachun — APPROVED** [review]

File: src/common/procedure/src/error.rs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    DuplicateProcedure {
+        procedure_id: ProcedureId,
+        backtrace: Backtrace,
+    },
+
+    // Currently we don't provide backtrace for an invalid ProcedureId as
+    // this error is mainly for the procedure manager to check whether a
+    // key in the procedure store is valid.
+    #[snafu(display("Invalid procedure id, source: {}", source))]
+    InvalidId { source: uuid::Error },
```
Comment (v0y4g3r) [inline src/common/procedure/src/error.rs:53]: Also add backtrace to `InvalidId`?

File: src/common/procedure/src/procedure.rs [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+// TODO(yingwen): Find a way to return the error message if the procedure is failed.
+/// State of a submitted procedure.
+#[derive(Debug)]
+pub enum ProcedureState {
+    /// The procedure is running.
+    Running,
+    /// The procedure is finished.
+    Done,
+    /// The procedure is failed.
+    Failed,
```
Comment (v0y4g3r) [inline src/common/procedure/src/procedure.rs:159]: Do we need to add a field to indicate if the failure is temporary (can be rerun) or permenant (should not rerun) like [opendal](https://docs.rs/opendal/0.25.1/opendal/struct.Error.html#method.is_temporary) does?

Comment (evenyag) [reply src/common/procedure/src/procedure.rs:159]: The failure should be permenant, I will explain this in the comment.

File: src/common/procedure/src/error.rs (hunk 4 above)
Comment (evenyag) [reply src/common/procedure/src/error.rs:53]: In fact, the backtrace is useless as the procedure manager usually uses this to test whether a path is valid and skip that path if it doesn't contain a procedure id.

I think I can define a dedicated error for this like RecvError of channel.

Comment (evenyag) [reply src/common/procedure/src/error.rs:53]: In fact, the backtrace is useless as the procedure manager usually uses this to test whether a path is valid and skip that path if it doesn't contain a procedure id.

I think I can define a dedicated error for this like RecvError of channel.

**Review by v0y4g3r — APPROVED** [review]
LGTM

**Pull request merged**